library/core/src/intrinsics.rs

```diff
@@ -2489,6 +2489,51 @@ extern "rust-intrinsic" {
     /// constructing an empty slice) is returned.
     #[rustc_nounwind]
     pub fn option_payload_ptr<T>(arg: *const Option<T>) -> *const T;
+
+    /// Returns whether the argument's value is statically known at
+    /// compile-time.
+    ///
+    /// This is useful when there is a way of writing the code that will
+    /// be *faster* when some variables have known values, but *slower*
+    /// in the general case: an `if is_val_statically_known(var)` can be used
+    /// to select between these two variants. The `if` will be optimized away
+    /// and only the desired branch remains.
+    ///
+    /// Formally speaking, this function non-deterministically returns `true`
+    /// or `false`, and the caller has to ensure sound behavior for both cases.
+    /// In other words, the following code has *Undefined Behavior*:
+    ///
+    /// ```rust
+    /// if !is_val_statically_known(0) { unreachable_unchecked(); }
+    /// ```
+    ///
+    /// This also means that the following code's behavior is unspecified; it
+    /// may panic, or it may not:
+    ///
+    /// ```rust,no_run
+    /// assert_eq!(is_val_statically_known(0), black_box(is_val_statically_known(0)))
+    /// ```
+    ///
+    /// Unsafe code may not rely on `is_val_statically_known` returning any
+    /// particular value, ever. However, the compiler will generally make it
+    /// return `true` only if the value of the argument is actually known.
+    ///
+    /// When calling this in a `const fn`, both paths must be semantically
+    /// equivalent, that is, the result of the `true` branch and the `false`
+    /// branch must return the same value and have the same side-effects *no
+    /// matter what*.
```

Comment:
Are they allowed to differ in panics? E.g. if an invalid input is handed to a safe function then it may either panic or return garbage and this differs between the branches?

Reply:
I'd say they must either both panic or neither panic. But the exact panic location is not relevant.

Reply:
That should be documented then. Though I think it might be tricky to uphold. Would it be possible to have `is_val_statically_known` to always stick to one value in const eval? That would be less footgun-ish.

Reply:
we could (in debug mode?) fork the const evaluator and run both paths and compare the result 🙃

Reply:
Does CTFE support that? If it does I'd definitely be in favor of doing this.

Reply:
not really, and it would be very expensive. It's ok. We're allowed to just run one path. I guess we could pick a different path depending on the even-ness of the minor compiler version or something funny or do something similar to -Zrandomize-layout

Reply:
> That should be documented then. Though I think it might be tricky to uphold. Would it be possible to have is_val_statically_known to always stick to one value in const eval? That would be less footgun-ish.

It's the same with `const_eval_select`. We already have an [RFC](https://github.com/rust-lang/rfcs/pull/3352) up for resolving this properly; in the mean time I think we should keep the strict rules.

Reply:
> They do in `pow`.

Could `#[track_caller]` and custom panic messages fix this? For that matter, is there a reason `pow` didn't have `#[track_caller]` to start out with?

Reply:
Don't think there's any reason not to. Custom panic messages would be a nice addition also

Reply:
I want to add, for the benefit of those reading in the future, that the core library already intentionally gives different panic messages depending on `const_eval_select`. It appears this is for performance reasons because the runtime variant is static, while the compile time variant is not. As far as I know, that is the *only* current use of `const_eval_select`.